### ISSUE TITLE
Add support for private repos on GitHub.com 

### ISSUE THREAD

**Issue by calebhailey** [sensu/bonsai#325]
Bonsai should support the same workflow for adding private repositories as it does for public repositories. 

Add an option in the "manage asset" menu to allow private repo maintainers to toggle visibility of the "Repository URL" link in the Bonsai sidebar. 

<img width="240" alt="Screen Shot 2021-10-11 at 2 21 18 PM" src="https://user-images.githubusercontent.com/141373/136857067-ffc77339-b0b0-4631-b668-e04ecba43293.png">

TODO: 

- [x] Update list of repos that are available to add to Bonsai (currently filtering out private repos)
- [ ] Annotate private repos in the list (e.g. `sensu/sensu-pagerduty-handler (private)`)
- [ ] Verify that fetching of GitHub release artifacts uses GitHub auth tokens (this will be required for private repos)

**Comment by calebhailey** [comment]
Updates: 

* Instead of (or in addition to) using an OAuth token
* Option 1: 
  - Prompt new users to choose public only or all repos
  - Add a button or instructions to [revoke the existing OAuth token](https://github.com/settings/applications) and redirect them to re-authenticate (where they can choose to grant access to all repos)
* Option 2: 
  - Default to public-only repos for new users 
  - Add a button or instructions to [revoke the existing OAuth token](https://github.com/settings/applications) and redirect them to a page where they can sign-up with a request for all repos
* We will need to scrub the text content in the app and user documentation that reference "public repositories"

**Comment by calebhailey** [comment]
@demillir let's see if Option 2 will work, as discussed on our call today. ☝️

**Comment by jspaleta** [comment]
Okay attempting to upload private repo and things appear to be wedged.
Here's how far I got:
1. logged into bonsai and was able to switch my oauth based access to allow full private repo access
2. selected the private repo `jspaleta/private-asset-test` (with no releases defined) and able to register it with bonsai
3. bonsai correctly indicated no releases found
4. added release 0.0.1 to private repository and filled release with files
5. used recompile button in bonsai web-ui
6. appears to be stuck at `Compiling Extension % 66.67 Complete`  for several minutes

I'm going to restart the staging environment now and restart to see if that clears things up

**Comment by jspaleta** [comment]
Okay I caught an error in the heroku application logs
```
2022-01-25T06:01:30.133274+00:00 app[worker.1]: 4 TID-owxypoed0 WARN: NoMethodError: undefined method `strip' for nil:NilClass

2022-01-25T06:01:30.133333+00:00 app[worker.1]: 4 TID-owxypoed0 WARN: /app/app/workers/sync_extension_contents_at_versions_worker.rb:252:in `set_commit_count'

2022-01-25T06:01:30.133334+00:00 app[worker.1]: /app/app/workers/sync_extension_contents_at_versions_worker.rb:59:in `sync_extension_version'

2022-01-25T06:01:30.133334+00:00 app[worker.1]: /app/app/workers/sync_extension_contents_at_versions_worker.rb:24:in `block in perform'

2022-01-25T06:01:30.133335+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/activerecord-5.2.2/lib/active_record/locking/pessimistic.rb:84:in `block in with_lock'

2022-01-25T06:01:30.133336+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/activerecord-5.2.2/lib/active_record/connection_adapters/abstract/database_statements.rb:259:in `block in transaction'

2022-01-25T06:01:30.133336+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/activerecord-5.2.2/lib/active_record/connection_adapters/abstract/transaction.rb:239:in `block in within_new_transaction'

2022-01-25T06:01:30.133337+00:00 app[worker.1]: /app/vendor/ruby-2.5.1/lib/ruby/2.5.0/monitor.rb:226:in `mon_synchronize'

2022-01-25T06:01:30.133337+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/activerecord-5.2.2/lib/active_record/connection_adapters/abstract/transaction.rb:236:in `within_new_transaction'

2022-01-25T06:01:30.133337+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/activerecord-5.2.2/lib/active_record/connection_adapters/abstract/database_statements.rb:259:in `transaction'

2022-01-25T06:01:30.133338+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/activerecord-5.2.2/lib/active_record/transactions.rb:212:in `transaction'

2022-01-25T06:01:30.133338+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/activerecord-5.2.2/lib/active_record/transactions.rb:301:in `transaction'

2022-01-25T06:01:30.133339+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/activerecord-5.2.2/lib/active_record/locking/pessimistic.rb:82:in `with_lock'

2022-01-25T06:01:30.133339+00:00 app[worker.1]: /app/app/workers/sync_extension_contents_at_versions_worker.rb:15:in `perform'

2022-01-25T06:01:30.133340+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/processor.rb:185:in `execute_job'

2022-01-25T06:01:30.133340+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/processor.rb:167:in `block (2 levels) in process'

2022-01-25T06:01:30.133341+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/middleware/chain.rb:128:in `block in invoke'

2022-01-25T06:01:30.133341+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-status-1.1.4/lib/sidekiq-status/server_middleware.rb:50:in `call'

2022-01-25T06:01:30.133342+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/middleware/chain.rb:130:in `block in invoke'

2022-01-25T06:01:30.133342+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/airbrake-10.0.1/lib/airbrake/sidekiq.rb:11:in `block in call'

2022-01-25T06:01:30.133343+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/airbrake-ruby-4.13.0/lib/airbrake-ruby/benchmark.rb:13:in `measure'

2022-01-25T06:01:30.133343+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/airbrake-10.0.1/lib/airbrake/sidekiq.rb:10:in `call'

2022-01-25T06:01:30.133344+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/middleware/chain.rb:130:in `block in invoke'

2022-01-25T06:01:30.133344+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/middleware/chain.rb:133:in `invoke'

2022-01-25T06:01:30.133344+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/processor.rb:166:in `block in process'

2022-01-25T06:01:30.133344+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/processor.rb:137:in `block (6 levels) in dispatch'

2022-01-25T06:01:30.133345+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/job_retry.rb:108:in `local'

2022-01-25T06:01:30.133345+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/processor.rb:136:in `block (5 levels) in dispatch'

2022-01-25T06:01:30.133355+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/rails.rb:43:in `block in call'

2022-01-25T06:01:30.133356+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.2/lib/active_support/execution_wrapper.rb:87:in `wrap'

2022-01-25T06:01:30.133359+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.2/lib/active_support/reloader.rb:73:in `block in wrap'

2022-01-25T06:01:30.133360+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.2/lib/active_support/execution_wrapper.rb:87:in `wrap'

2022-01-25T06:01:30.133360+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.2/lib/active_support/reloader.rb:72:in `wrap'

2022-01-25T06:01:30.133360+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/rails.rb:42:in `call'

2022-01-25T06:01:30.133360+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/processor.rb:132:in `block (4 levels) in dispatch'

2022-01-25T06:01:30.133361+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/processor.rb:243:in `stats'

2022-01-25T06:01:30.133361+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/processor.rb:127:in `block (3 levels) in dispatch'

2022-01-25T06:01:30.133361+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/job_logger.rb:8:in `call'

2022-01-25T06:01:30.133362+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/processor.rb:126:in `block (2 levels) in dispatch'

2022-01-25T06:01:30.133364+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/job_retry.rb:73:in `global'

2022-01-25T06:01:30.133365+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/processor.rb:125:in `block in dispatch'

2022-01-25T06:01:30.133365+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/logging.rb:48:in `with_context'

2022-01-25T06:01:30.133365+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/logging.rb:42:in `with_job_hash_context'

2022-01-25T06:01:30.133365+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/processor.rb:124:in `dispatch'

2022-01-25T06:01:30.133366+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/processor.rb:165:in `process'

2022-01-25T06:01:30.133366+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/processor.rb:83:in `process_one'

2022-01-25T06:01:30.133366+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/processor.rb:71:in `run'

2022-01-25T06:01:30.133367+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/util.rb:16:in `watchdog'

2022-01-25T06:01:30.133367+00:00 app[worker.1]: /app/vendor/bundle/ruby/2.5.0/gems/sidekiq-5.2.5/lib/sidekiq/util.rb:25:in `block in safe_thread'

2022-01-25T06:01:41.705091+00:00 heroku[router]: at=info method=GET path="/assets/jspaleta/107/sync_status?_=1643090489825" host=bonsai-asset-index-staging.herokuapp.com request_id=6599a52d-d7f6-4a7b-9b21-3a158459961c fwd="72.42.136.178" dyno=web.1 connect=0ms service=28ms status=200 bytes=1336 protocol=https

```

**Comment by jspaleta** [comment]
Quick check just to make sure public repo registering in the staging environment still works.. The error above is definitely related to my private repo... not sure what expectations the private repo api response is breaking.

**Comment by mbiang** [comment]
This is due to the [sync_extension_version](https://github.com/sensu/bonsai/blob/master/app/workers/sync_extension_contents_at_versions_worker.rb#L52) method of `SyncExtensionContentsAtVersionsWorker` attempting to checkout a local copy of the Git Repository.  This does not work with private assets, as `git checkout` will fail because it is not authenticated and cannot access the private repository.

We either need to:
- See if we can log into authenticate with the git CLI using a Github Oauth Token, or
- Refactor local git methods to instead use the Github API

**Comment by jspaleta** [comment]
Update:

Looks like private repos successfully compile now on staging using deployed fix for #348.
